File: JS/nodejs/test/certify.js
// SPDX-License-Identifier: LGPL-3.0-only
'use strict';

// Certifies src/frame.js byte-for-byte against the cross-language golden vectors in
// Documentation/golden_vectors.json. Dependency-free — stdlib `fs`/`path` + JSON.parse.
//   node JS/nodejs/test/certify.js   (from the repo root)
//   npm --prefix JS/nodejs run certify

const fs = require('fs');
const path = require('path');
const { crc16, encode, decode, syndrome, FRAME_SIZE, SYNC, CRC_COVER, BROADCAST } = require('../src/frame.js');

function loadGolden() {
  const candidates = [
    path.join(__dirname, '..', '..', '..', 'Documentation', 'golden_vectors.json'),
    path.join(__dirname, '..', '..', '..', 'python', 'MCP', 'golden_vectors.json'),
  ];
  for (const p of candidates) {
    if (fs.existsSync(p)) return JSON.parse(fs.readFileSync(p, 'utf8'));
  }
  throw new Error('golden_vectors.json not found in expected locations');
}

let fails = 0;
function check(cond, label) {
  console.log((cond ? '  PASS  ' : '  FAIL  ') + label);
  if (!cond) fails += 1;
}

function rawValid(b) {
  if (b.length !== FRAME_SIZE || (b[0] & 0xFF) !== SYNC) return false;
  const stored = ((b[15] & 0xFF) << 8) | (b[16] & 0xFF);
  return crc16(b, CRC_COVER) === stored;
}

const gv = loadGolden();

// 1. CRC anchors
const c1 = crc16(Buffer.from('123456789', 'ascii'));
check(c1 === 0x29B1, `CRC("123456789") = 0x${c1.toString(16)} (want 0x29b1)`);
const c0 = crc16(Buffer.alloc(15));
check(c0 === 0x4EC3, `CRC(0^15) = 0x${c0.toString(16)} (want 0x4ec3)`);

// 2. example frame anchor
const exGot = encode({
  version: 1, type: 3, seq: 0x1234, src: 1, dst: BROADCAST,
  payload: Buffer.from([0xDE, 0xAD, 0xBE, 0xEF]), tsUs: 0xAB12CD,
}).toString('hex');
check(exGot === gv.anchors.exampleFrame_full.toLowerCase(), `exampleFrame_full: got ${exGot}`);

// 3. encode_basis: raw-CRC-valid + (known types) decode/roundtrip
check(gv.encode_basis.length > 0, 'encode_basis vectors present');
let encFails = 0;
gv.encode_basis.forEach((v, i) => {
  const raw = Buffer.from(v.frame, 'hex');
  if (!rawValid(raw)) {
    encFails += 1;
    console.log(`  FAIL  encode_basis[${i}]: raw CRC invalid`);
    return;
  }
  if ((raw[1] & 0x0F) <= 3) {
    try {
      const f = decode(raw);
      if (!encode(f).equals(raw)) {
        encFails += 1;
        console.log(`  FAIL  encode_basis[${i}]: roundtrip mismatch`);
      }
    } catch (e) {
      encFails += 1;
      console.log(`  FAIL  encode_basis[${i}]: ${e.message}`);
    }
  }
});
fails += encFails;
if (encFails === 0) console.log(`  PASS  ${gv.encode_basis.length} encode_basis vectors (decode + roundtrip)`);

// 4. syndrome_basis: reproduce each basis word and check its syndrome
check(gv.syndrome_basis.length > 0, 'syndrome_basis vectors present');
let synFails = 0;
gv.syndrome_basis.forEach((v, i) => {
  const word = Buffer.alloc(FRAME_SIZE);
  if (typeof v.bit === 'number') {
    word[Math.floor(v.bit / 8)] = 1 << (7 - (v.bit % 8));
  } else if (typeof v.word === 'string') {
    Buffer.from(v.word, 'hex').copy(word);
  }
  const got = syndrome(word);
  if (got !== v.syndrome) {
    synFails += 1;
    console.log(`  FAIL  syndrome_basis[${i}]: got 0x${got.toString(16)} want 0x${v.syndrome.toString(16)}`);
  }
});
fails += synFails;
if (synFails === 0) console.log(`  PASS  ${gv.syndrome_basis.length} syndrome_basis vectors`);

console.log('');
if (fails === 0) {
  console.log(`ALL CHECKS PASSED — Node.js codec cemented (${gv.encode_basis.length} encode + ${gv.syndrome_basis.length} syndrome).`);
  process.exit(0);
} else {
  console.log(`${fails} certification check(s) FAILED`);
  process.exit(1);
}

File: haskell/test/Certify.hs
-- SPDX-License-Identifier: LGPL-3.0-only
{-# LANGUAGE OverloadedStrings #-}

-- | Certifies the Haskell DeModFrame codec ("DCF.Transport.FrameSpec") byte-for-byte
-- against the cross-language golden vectors in Documentation/golden_vectors.json.
-- Exits non-zero if any anchor, encode-basis, or syndrome-basis check fails.
module Main (main) where

import           Control.Monad        (forM_, unless, when)
import           Data.Aeson           (Value (..), decode)
import           Data.Aeson.Key       (Key)
import qualified Data.Aeson.KeyMap    as KM
import           Data.Bits            (shiftL, xor, (.&.), (.|.))
import qualified Data.ByteString.Lazy as BL
import           Data.Char            (digitToInt, intToDigit)
import           Data.Foldable        (toList)
import           Data.IORef           (modifyIORef', newIORef, readIORef)
import           Data.Maybe           (fromMaybe)
import           Data.Scientific      (Scientific, toBoundedInteger)
import qualified Data.Text            as T
import           Data.Word            (Word16, Word8)
import           Numeric              (showHex)
import           System.Directory     (doesFileExist)
import           System.Exit          (exitFailure, exitSuccess)

import           DCF.Transport.FrameSpec
import           DCF.Transport.SuperPack (isSuperPack, packSuper, unpackSuper)

main :: IO ()
main = do
  path <- findGolden
  raw  <- BL.readFile path
  let root = fromMaybe (error ("cannot parse JSON: " ++ path)) (decode raw :: Maybe Value)
  fails <- newIORef (0 :: Int)
  let bad msg = modifyIORef' fails (+ 1) >> putStrLn ("  FAIL  " ++ msg)
      ok  msg = putStrLn ("  PASS  " ++ msg)

  -- 1. CRC anchors
  let crc1 = crc16ccitt (strBytes "123456789")
  if crc1 == 0x29B1 then ok "CRC(\"123456789\") = 0x29B1"
                    else bad ("CRC(\"123456789\") = 0x" ++ hex16 crc1)
  let crc0 = crc16ccitt (replicate 15 0)
  if crc0 == 0x4EC3 then ok "CRC(0^15) = 0x4EC3"
                    else bad ("CRC(0^15) = 0x" ++ hex16 crc0)

  -- 2. example frame anchor
  let exHex   = T.unpack (str (look "exampleFrame_full" (look "anchors" root)))
      exFrame = Frame 1 Ctrl 0x1234 1 broadcast (0xDE, 0xAD, 0xBE, 0xEF) 0xAB12CD
      exGot   = bytesHex (encodeFrame exFrame)
  if exGot == exHex then ok "exampleFrame_full matches"
                    else bad ("exampleFrame: got " ++ exGot ++ " want " ++ exHex)

  -- 3. encode_basis: raw-CRC-valid + (known types) decode/roundtrip
  let encs = arr (look "encode_basis" root)
  forM_ (zip [0 :: Int ..] encs) $ \(i, o) -> do
    let bytes = hexBytes (T.unpack (str (look "frame" o)))
    if not (rawValid bytes)
      then bad ("encode_basis[" ++ show i ++ "]: raw CRC invalid")
      else when ((bytes !! 1) .&. 0x0F <= 3) $
             case decodeFrame bytes of
               Right f  -> unless (encodeFrame f == bytes) $
                             bad ("encode_basis[" ++ show i ++ "]: roundtrip mismatch")
               Left err -> bad ("encode_basis[" ++ show i ++ "]: " ++ show err)

  -- 4. syndrome_basis
  let syns = arr (look "syndrome_basis" root)
  forM_ (zip [0 :: Int ..] syns) $ \(i, o) -> do
    let word = case lookMaybe "bit" o of
                 Just (Number n) -> oneBit (sciInt n)
                 _               -> replicate frameSize 0
        expected = fromIntegral (numInt (look "syndrome" o)) :: Word16
        got = crc16ccitt (take 15 word) `xor` makeW16 (word !! 15) (word !! 16)
    when (got /= expected) $
      bad ("syndrome_basis[" ++ show i ++ "]: got 0x" ++ hex16 got ++ " want 0x" ++ hex16 expected)

  -- 5. SuperPack container vectors (Documentation/superpack_vectors.json)
  spPath <- findSuper
  spRaw  <- BL.readFile spPath
  let spRoot  = fromMaybe (error ("cannot parse JSON: " ++ spPath)) (decode spRaw :: Maybe Value)
      spCases = arr (look "cases" spRoot)
  forM_ (zip [0 :: Int ..] spCases) $ \(i, o) -> do
    let aHex = T.unpack (str (look "a" o))
        bHex = T.unpack (str (look "b" o))
        spHx = T.unpack (str (look "super" o))
    case packSuper (hexBytes aHex) (hexBytes bHex) of
      Left e   -> bad ("superpack[" ++ show i ++ "]: pack " ++ e)
      Right pk -> do
        unless (isSuperPack pk) $ bad ("superpack[" ++ show i ++ "]: not recognised")
        unless (bytesHex pk == spHx) $ bad ("superpack[" ++ show i ++ "]: pack bytes")
    case unpackSuper (hexBytes spHx) of
      Left e         -> bad ("superpack[" ++ show i ++ "]: unpack " ++ e)
      Right (ra, rb) ->
        unless (bytesHex ra == aHex && bytesHex rb == bHex) $
          bad ("superpack[" ++ show i ++ "]: unpack mismatch")
  let zeroFrame = encodeFrame (Frame 1 Data 0 0 0 (0, 0, 0, 0) 0)
  case packSuper zeroFrame zeroFrame of
    Right spz -> unless (makeW16 (spz !! 30) (spz !! 31) == 0x5B75) $
                   bad "superpack zero-core anchor /= 0x5B75"
    Left e    -> bad ("superpack zero-core: " ++ e)

  n <- readIORef fails
  putStrLn ""
  if n == 0
    then do putStrLn ("ALL CHECKS PASSED (" ++ show (length encs) ++ " encode + "
                      ++ show (length syns) ++ " syndrome + " ++ show (length spCases)
                      ++ " superpack) — Haskell codec cemented.")
            exitSuccess
    else do putStrLn (show n ++ " certification check(s) FAILED")
            exitFailure

-- ── JSON helpers ────────────────────────────────────────────────────────────

findGolden :: IO FilePath
findGolden = go [ "../Documentation/golden_vectors.json"
                , "Documentation/golden_vectors.json"
                , "../python/MCP/golden_vectors.json" ]
  where
    go []       = error "golden_vectors.json not found in expected locations"
    go (p : ps) = do e <- doesFileExist p; if e then return p else go ps

findSuper :: IO FilePath
findSuper = go [ "../Documentation/superpack_vectors.json"
               , "Documentation/superpack_vectors.json"
               , "../python/MCP/superpack_vectors.json" ]
  where
    go []       = error "superpack_vectors.json not found (run gen_superpack_vectors.py)"
    go (p : ps) = do e <- doesFileExist p; if e then return p else go ps

look :: Key -> Value -> Value
look k (Object o) = fromMaybe (error ("missing key " ++ show k)) (KM.lookup k o)
look k _          = error ("expected object for key " ++ show k)

lookMaybe :: Key -> Value -> Maybe Value
lookMaybe k (Object o) = KM.lookup k o
lookMaybe _ _          = Nothing

arr :: Value -> [Value]
arr (Array a) = toList a
arr _         = error "expected array"

str :: Value -> T.Text
str (String t) = t
str _          = error "expected string"

numInt :: Value -> Int
numInt (Number n) = sciInt n
numInt _          = error "expected number"

sciInt :: Scientific -> Int
sciInt s = fromMaybe (error "non-integer number") (toBoundedInteger s)

-- ── byte / hex helpers ──────────────────────────────────────────────────────

strBytes :: String -> [Word8]
strBytes = map (fromIntegral . fromEnum)

hexBytes :: String -> [Word8]
hexBytes []           = []
hexBytes (a : b : cs) = fromIntegral (digitToInt a * 16 + digitToInt b) : hexBytes cs
hexBytes _            = error "odd-length hex string"

bytesHex :: [Word8] -> String
bytesHex = concatMap byteHex
  where
    byteHex w = [ intToDigit (fromIntegral w `div` 16)
                , intToDigit (fromIntegral w `mod` 16) ]

hex16 :: Word16 -> String
hex16 w = showHex w ""

makeW16 :: Word8 -> Word8 -> Word16
makeW16 h l = fromIntegral h `shiftL` 8 .|. fromIntegral l

oneBit :: Int -> [Word8]
oneBit bit = [ if i == bit `div` 8 then 1 `shiftL` (7 - bit `mod` 8) else 0
             | i <- [0 .. frameSize - 1] ]

rawValid :: [Word8] -> Bool
rawValid bytes =
     length bytes == frameSize
  && head bytes == syncByte
  && crc16ccitt (take 15 bytes) == makeW16 (bytes !! 15) (bytes !! 16)
